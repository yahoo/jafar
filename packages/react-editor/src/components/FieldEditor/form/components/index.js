/**
  * Copyright 2020, Verizon Media
  * Licensed under the terms of the MIT license. See LICENSE file in project root for terms.
  */

import TextInput from '@jafar/react-components/edit/Text';
import Checkbox from '@jafar/react-components/edit/Checkbox';
import MultiSelectFields from './MultiSelectFields';
import Validators from './Validators';
import Handler from './Handler';
import Term from './Term';

export default {
  TextInput,
  Checkbox,
  MultiSelectFields,
  Validators,
  Handler,
  Term,
};
