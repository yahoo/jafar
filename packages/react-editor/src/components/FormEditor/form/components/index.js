/**
  * Copyright 2020, Verizon Media
  * Licensed under the terms of the MIT license. See LICENSE file in project root for terms.
  */

import TextInput from '@jafar/react-components/edit/Text';
import NumberInput from '@jafar/react-components/edit/Number';
import JsonEditor from '@jafar/react-components/edit/JsonEditor';
import Fields from './Fields';

export default {
  TextInput,
  NumberInput,
  JsonEditor,
  Fields,
};
