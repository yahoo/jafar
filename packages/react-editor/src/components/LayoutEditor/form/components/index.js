/**
  * Copyright 2020, Verizon Media
  * Licensed under the terms of the MIT license. See LICENSE file in project root for terms.
  */

import TextInput from '@jafar/react-components/edit/Text';
import NumberInput from '@jafar/react-components/edit/Number';
import Select from '@jafar/react-components/edit/Select';
import DynamicItemsModal from '@jafar/react-components/edit/DynamicItemsModal';

export default {
  TextInput,
  NumberInput,
  Select,
  DynamicItemsModal,
};
