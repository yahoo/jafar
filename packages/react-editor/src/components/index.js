/**
  * Copyright 2020, Verizon Media
  * Licensed under the terms of the MIT license. See LICENSE file in project root for terms.
  */

import FieldEditor from './FieldEditor';
import FormEditor from './FormEditor';

export {
  FieldEditor,
  FormEditor,
};

