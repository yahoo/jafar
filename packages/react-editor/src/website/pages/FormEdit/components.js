/**
  * Copyright 2020, Verizon Media
  * Licensed under the terms of the MIT license. See LICENSE file in project root for terms.
  */

import _TextInput from '@jafar/react-components/edit/Text';
import _NumberInput from '@jafar/react-components/edit/Number';
import _Checkbox from '@jafar/react-components/edit/Checkbox';
import _CheckboxCollection from '@jafar/react-components/edit/CheckboxCollection';
import _Select from '@jafar/react-components/edit/Select';
import _MultiSelect from '@jafar/react-components/edit/MultiSelect';
import _JsonEditor from '@jafar/react-components/edit/JsonEditor';
import _Switch from '@jafar/react-components/edit/Switch';
import _Url from '@jafar/react-components/view/Url';
import _Boolean from '@jafar/react-components/view/Boolean';
import _JsonView from '@jafar/react-components/view/JsonView';

export const TextInput = _TextInput;
export const NumberInput = _NumberInput;
export const Checkbox = _Checkbox;
export const CheckboxCollection = _CheckboxCollection;
export const Select = _Select;
export const MultiSelect = _MultiSelect;
export const JsonEditor = _JsonEditor;
export const Switch = _Switch;
export const Url = _Url;
export const JsonView = _JsonView;
export const Boolean = _Boolean;

// contains only common components of this app that any entity can use
export default {
  TextInput,
  NumberInput,
  Checkbox,
  CheckboxCollection,
  Select,
  MultiSelect,
  JsonEditor,
  Url,
  Switch,
  JsonView,
  Boolean,
};
