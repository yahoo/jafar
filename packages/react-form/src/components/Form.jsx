/**
  * Copyright 2020, Verizon Media
  * Licensed under the terms of the MIT license. See LICENSE file in project root for terms.
  */

import React from 'react';
import PropTypes from 'prop-types';
import JForm from '@jafar/form';
import FormContext from './context';

export default class Form extends React.Component {
  static contextType = FormContext;

  static propTypes = {
    model: PropTypes.object.isRequired,
    resources: PropTypes.object,
    settings: PropTypes.object,
    data: PropTypes.object,
    context: PropTypes.object,
  };

  constructor(props) {
    super(props);
    // stored to compare pointer changes for re-init / other actions (react don't pass prevState in getDerivedStateFromProps)
    this.state = {
      propsModel: props.model, 
      propsContext: props.context,
      propsData: props.data,
    };
  }

  static getDerivedStateFromProps(props, state) {
    // if model was changed (checking only for new pointer) - trigger init action
    if (props.model !== state.propsModel) {
      return ({ propsModel: props.model, needAction: 'init', destroyPrevModelId: state.propsModel.id });
    }
    // if context was changed (checking only for new pointer) - trigger changeContext action
    if (props.context !== state.propsContext) {
      return ({ propsContext: props.context, needAction: 'changeContext' });
    }
    // if data was changed (checking only for new pointer) - trigger changeData action
    if (props.data !== state.propsData) {
      return ({ propsData: props.data, needAction: 'changeData' });
    }
    return null;
  }

  render() {
    if (!this.state.form) {
      return null;
    }

    const context = {
      model: this.state.form.model,
      resources: this.state.form.resources,
      actions: {
        changeValue: this.privateForm.changeValue.bind(this.privateForm),
        changeData: this.privateForm.changeData.bind(this.privateForm),
        changeContext: this.privateForm.changeContext.bind(this.privateForm),
        changeState: this.privateForm.changeState.bind(this.privateForm),
        changeUi: this.privateForm.changeUi.bind(this.privateForm),
        submit: this.privateForm.submit.bind(this.privateForm),
        reset: this.privateForm.reset.bind(this.privateForm),
      },
      parent: this.context,
    };

    return (
      <FormContext.Provider value={context}>
        <React.Fragment>{this.props.children}</React.Fragment>
      </FormContext.Provider>);
  }

  componentDidMount() {
    this.init();
  }

  componentDidUpdate() {
    switch (this.state.needAction) {
    case 'init': {
      if (this.state.destroyPrevModelId !== this.state.propsModel.id) {
        this.privateForm.destroy();
      }
      this.init();
      this.setState({ needAction: undefined, destroyPrevModelId: undefined });
      break;
    }
    case 'changeContext': {
      this.privateForm.changeContext(this.props.context);
      this.setState({ needAction: undefined });
      break;
    }
    case 'changeData': {
      this.privateForm.changeData(this.props.data);
      this.setState({ needAction: undefined });
      break;
    }
    default: {}
    }
  }

  init = () => {
    this.privateForm = new JForm();
    const contextOverride = this.props.context ? { context: this.props.context } : undefined;
    const model = Object.assign({}, this.props.model, contextOverride);
    this.privateForm.init(model, this.props.resources, this.props.settings, this.updatePublicForm);
  }

  componentWillUnmount() {
    this.privateForm.destroy();
    this.destroyed = true;
  }

  updatePublicForm = (form) => {
    if (this.destroyed) return;

    this.setState({ form });
  };
}
