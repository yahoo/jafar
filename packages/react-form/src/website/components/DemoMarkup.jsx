/**
  * Copyright 2020, Verizon Media
  * Licensed under the terms of the MIT license. See LICENSE file in project root for terms.
  */

import React from 'react';
import PrismCode from 'react-prism';
import JSZip from 'jszip';
import 'prismjs';
import 'prismjs/themes/prism.css';
import Styled from './StyledComponents';

export default class DemoMarkup extends React.Component {
  constructor(props) {
    super(props);
    this.formFolder = this.props.formFolder || 'form';
    this.index = this.getFile('index');
    this.fields = this.getFile('fields');
    this.components = this.getFile('components');
    this.data = this.props.data ? this.getFile('data') : undefined;
    this.conversions = this.props.conversions ? this.getFile('conversions') : undefined;
    this.validators = this.props.validators ? this.getFile('validators') : undefined;
    this.terms = this.props.terms ? this.getFile('terms') : undefined;
    this.dependenciesChanges = this.props.dependenciesChanges ? this.getFile('dependencies-changes') : undefined;
    this.hooks = this.props.hooks ? this.getFile('hooks') : undefined;
    this.mockService = this.props.mockService ? this.getFile('mocks/service') : undefined;
    this.extraComponents = this.props.extraComponents ? this.props.extraComponents
      .map(filePath => this.getFileByFullRelativePath(filePath)) : undefined;
  }

  getFile(fileName) {
    return require(`!!raw-loader!../demos/react/${this.props.exampleName}/${this.formFolder}/${fileName}.js`).default;
  }

  getFileByFullRelativePath(filePath) {
    return require(`!!raw-loader!../demos/react/${this.props.exampleName}/${filePath}.jsx`).default;
  }

  download = () => {
    const rootFolder = new JSZip();

    // add form folder
    const formFolder = rootFolder.folder(this.formFolder);
    formFolder.file('index.js', this.index);
    formFolder.file('fields.js', this.fields);
    formFolder.file('components.js', this.components);
    if (this.data) formFolder.file('data.js', this.data);
    if (this.conversions) formFolder.file('conversions.js', this.conversions);
    if (this.validators) formFolder.file('validators.js', this.validators);
    if (this.terms) formFolder.file('terms.js', this.terms);
    if (this.dependenciesChanges) formFolder.file('dependenciesChanges.js', this.dependenciesChanges);
    if (this.hooks) formFolder.file('hooks.js', this.hooks);
    if (this.mockService) {
      const mocksFolder = formFolder.folder('mocks');
      mocksFolder.file('service.js', this.mockService);
    }
    if (this.extraComponents) {
      this.props.extraComponents.forEach((filePath, index) => {
        let file = filePath;
        let containerFolder = rootFolder;
        if (filePath.includes('/')) {
          const arr = filePath.split('/');
          file = arr[1];
          containerFolder = rootFolder.folder(arr[0]);
        }
        containerFolder.file(`${file}.jsx`, this.extraComponents[index]);
      });
    }

    // add demo file
    rootFolder.file('demo.jsx', this.props.demo);

    rootFolder.generateAsync({ type:'blob' }).then((content) => {
      const href = window.URL.createObjectURL(content);
      const link = document.createElement('a');
      link.setAttribute('href', href);
      link.setAttribute('download', `${this.props.exampleName}.zip`);
      link.style.display = 'none';
      document.body.appendChild(link);
      link.click();
      document.body.removeChild(link);
    });
  }

  render() {
    return (<Styled.DemoWrapper>
      <Styled.DownloadDemo onClick={this.download}>Download</Styled.DownloadDemo>
      <Styled.H3>Form definition</Styled.H3>
      <p>{this.formFolder}/fields.js</p>
      <PrismCode className="language-javascript" component="pre">{this.fields}</PrismCode>
      {
        this.data && <div>
          <p>{this.formFolder}/data.js</p>
          <PrismCode className="language-javascript" component="pre">{this.data}</PrismCode>
        </div>
      }
      <p>{this.formFolder}/components.js</p>
      <PrismCode className="language-javascript" component="pre">{this.components}</PrismCode>
      {
        this.props.conversions && <div>
          <p>{this.formFolder}/conversions.js</p>
          <PrismCode className="language-javascript" component="pre">{this.conversions}</PrismCode>
        </div>
      }
      {
        this.props.validators && <div>
          <p>{this.formFolder}/validators.js</p>
          <PrismCode className="language-javascript" component="pre">{this.validators}</PrismCode>
        </div>
      }
      {
        this.props.terms && <div>
          <p>{this.formFolder}/terms.js</p>
          <PrismCode className="language-javascript" component="pre">{this.terms}</PrismCode>
        </div>
      }
      {
        this.props.dependenciesChanges && <div>
          <p>{this.formFolder}/dependencies-changes.js</p>
          <PrismCode className="language-javascript" component="pre">{this.dependenciesChanges}</PrismCode>
        </div>
      }
      {
        this.props.hooks && <div>
          <p>{this.formFolder}/hooks.js</p>
          <PrismCode className="language-javascript" component="pre">{this.hooks}</PrismCode>
        </div>
      }
      <p>{this.formFolder}/index.js</p>
      <PrismCode className="language-javascript" component="pre">{this.index}</PrismCode>
      {
        this.props.mockService && <div>
          <p>{this.formFolder}/mocks/service.js</p>
          <PrismCode className="language-javascript" component="pre">{this.mockService}</PrismCode>
        </div>
      }
      {
        !this.props.hideHtml
        && <div>
          <Styled.H3>Components</Styled.H3>
          <p>Demo.jsx</p>
          <PrismCode className="language-javascript" component="pre">{this.props.demo}</PrismCode>
          {
            this.extraComponents && <div>{
              this.extraComponents.map((componentHtml, index) => (<div key={index}>
                <p>{this.props.extraComponents[index]}.jsx</p>
                <PrismCode className="language-javascript" component="pre">{componentHtml}</PrismCode>
              </div>))
            }</div>
          }
        </div>
      }
    </Styled.DemoWrapper>);
  }
}
