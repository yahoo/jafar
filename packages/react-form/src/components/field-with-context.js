/**
  * Copyright 2020, Verizon Media
  * Licensed under the terms of the MIT license. See LICENSE file in project root for terms.
  */

import React from 'react';
import PropTypes from 'prop-types';
import { mapFieldToProps } from '@jafar/form';
import FormContext from './context';

const propTypes = {
  id: PropTypes.string.isRequired,
};

export default function withContext(FieldView) {
  class Field extends React.Component {
    static contextType = FormContext;

    static propTypes = propTypes;

    render() {
      const fieldViewProps = mapFieldToProps(this.props.id, this.context.model, this.context.resources);

      return (
        <FieldView
          {...this.props} // custom props to propagate
          {...fieldViewProps}
          onValueChange={this.onValueChange}
          onStateChange={this.onStateChange}
        />
      );
    }

    // We don't pass 'changeValue' and 'changeState' directly to FieldView prop because context 
    // create a new instance of them each render causing un-necessary renders to FieldView
    onValueChange = (id, value) => {
      this.context.actions.changeValue(id, value);
    }

    onStateChange = (id, state) => {
      this.context.actions.changeState(id, state);
    }
  }

  return Field;
}
