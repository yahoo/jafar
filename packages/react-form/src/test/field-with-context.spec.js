/**
  * Copyright 2020, Verizon Media
  * Licensed under the terms of the MIT license. See LICENSE file in project root for terms.
  */

import React from 'react';
import PropTypes from 'prop-types';
import ShallowRenderer from 'react-test-renderer/shallow';
import { shallow } from 'enzyme';
import { withContext, FieldView } from '../components';

describe('field-with-context', () => {
  let Field;
  let context;

  beforeEach(() => {
    Field = withContext(FieldView);
    Field.contextTypes = {
      actions: PropTypes.object,
      model: PropTypes.object,
      resources: PropTypes.object,
    };

    context = {
      actions: {
        changeValue: jest.fn(),
        changeState: jest.fn(),
      },
      model: {
        fields: {
          firstName: {
            path: 'firstName',
            label: 'First Name',
            description: 'Enter first name',
            component: {
              name: 'InputText',
              value: 'Rachel',
            },
            invalid: false,
            errors: [],
          },
        },
      },
      resources: {
        components: {
          InputText: { renderer: jest.fn() },
        },
      },
    };
  });
  it('rendered correctly', () => {
    const element = new ShallowRenderer().render(<Field id="firstName" />, context);
    expect(element).toMatchSnapshot();
  });

  it('rendered correctly with unsupported jafar field props - pass it to the underline component', () => {
    const element = new ShallowRenderer().render(<Field id="firstName" mickey="mouse" />, context);
    expect(element).toMatchSnapshot();
  });

  it(`rendered correctly with reserved jafar field view prop - `
    + `dont pass it to the underline component, pass the jafar props`, () => {
    const element = new ShallowRenderer().render(<Field id="firstName" value="mouse" />, context);
    expect(element).toMatchSnapshot();
  });

  it(`rendered correctly with reserved jafar field view prop that is undefined - `
    + `dont pass it to the underline component, pass the jafar props`, () => {
    context.model.fields.firstName.component.value = undefined;
    const element = new ShallowRenderer().render(<Field id="firstName" value="mouse" />, context);
    expect(element).toMatchSnapshot();
  });

  it('onValueChange calls context.changeValue with correct args', () => {
    const element = new shallow(<Field id="firstName" />);
    element.instance().context = context;
    element.instance().onValueChange('firstName', 'new value');
    expect(context.actions.changeValue).toHaveBeenCalledWith('firstName', 'new value');
  });

  it('onStateChange calls context.changeState with correct args', () => {
    const element = new shallow(<Field id="firstName" />);
    element.instance().context = context;
    element.instance().onStateChange('firstName', { a: 'new value' });
    expect(context.actions.changeState).toHaveBeenCalledWith('firstName', { a: 'new value' });
  });
});
