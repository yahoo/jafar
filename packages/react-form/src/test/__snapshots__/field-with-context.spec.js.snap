// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`field-with-context rendered correctly 1`] = `
<FieldView
  component={[MockFunction]}
  description="Enter first name"
  dirty={false}
  disabled={false}
  empty={false}
  errors={Array []}
  excluded={false}
  id="firstName"
  invalid={false}
  label="First Name"
  onStateChange={[Function]}
  onValueChange={[Function]}
  required={false}
  state={Object {}}
  value="Rachel"
/>
`;

exports[`field-with-context rendered correctly with reserved jafar field view prop - dont pass it to the underline component, pass the jafar props 1`] = `
<FieldView
  component={[MockFunction]}
  description="Enter first name"
  dirty={false}
  disabled={false}
  empty={false}
  errors={Array []}
  excluded={false}
  id="firstName"
  invalid={false}
  label="First Name"
  onStateChange={[Function]}
  onValueChange={[Function]}
  required={false}
  state={Object {}}
  value="Rachel"
/>
`;

exports[`field-with-context rendered correctly with reserved jafar field view prop that is undefined - dont pass it to the underline component, pass the jafar props 1`] = `
<FieldView
  component={[MockFunction]}
  description="Enter first name"
  dirty={false}
  disabled={false}
  empty={false}
  errors={Array []}
  excluded={false}
  id="firstName"
  invalid={false}
  label="First Name"
  onStateChange={[Function]}
  onValueChange={[Function]}
  required={false}
  state={Object {}}
/>
`;

exports[`field-with-context rendered correctly with unsupported jafar field props - pass it to the underline component 1`] = `
<FieldView
  component={[MockFunction]}
  description="Enter first name"
  dirty={false}
  disabled={false}
  empty={false}
  errors={Array []}
  excluded={false}
  id="firstName"
  invalid={false}
  label="First Name"
  mickey="mouse"
  onStateChange={[Function]}
  onValueChange={[Function]}
  required={false}
  state={Object {}}
  value="Rachel"
/>
`;
